[nativert] Move Pytree

yushangdi opened the pull request:
Summary: fbcode/sigmoid/core/common -> fbcode/caffe2/torch/nativert/common

Torch Native Runtime RFC: https://github.com/pytorch/rfcs/pull/72

Test Plan:
```
buck run fbcode//mode/dev-nosan  //caffe2/test/cpp/nativert:pytree_test
```

OSS CI

Rollback Plan:

Differential Revision: D75965059

Skylion007 commented:
I'm sure @XuehaiPan has some thoughts here

yushangdi commented:
> Also this pytree impl contains many optimizations we accumulated to high qps low latency model serving use cases and imo we should still recommend ppl to use optree in common cases.

@zhxchen17 sure, I'll move it to `nativert/detail` instead then.

albanD reviewed (dismissed):
@zou3519 how does that align with the other c++ pytree discussion we've been having?

zou3519 commented:
> @zou3519 how does that align with the other c++ pytree discussion we've been having?

@albanD when we say C++ pytree we really mean two things:
1) Python bindings for C++ functionality to flatten/unflatten PyObject* Python containers
2) C++ functionality to flatten/unflatten IValues

"C++ Pytree" and optree is (1). We don't have (2) in PyTorch. This PR adds (2).
There are use cases for both, e.g. executorch has both (1) and (2).

My conclusion is that we should allow this PR and that it is unrelated to the optree story. If we're up to it then we should probably name it something else other than "pytree", because there's no Python involved

yushangdi commented:
>  If we're up to it then we should probably name it something else other than "pytree", because there's no Python involved

I'd argue it's not completely unrelated to python, because it does register and process the string representation of python types, e.g. `builtins.tuple`, `torch.fx.immutable_collections.immutable_dict`...

But if we do want to change to a name other than pytree, @zou3519 @zhxchen17  do you have any suggestion for a better name?

XuehaiPan commented on torch/nativert/detail/Pytree.h:
```diff
@@ -0,0 +1,168 @@
+
+class TreeSpec;
+
+using TreeFlattenFn =
+    void (*)(const c10::IValue&, const TreeSpec&, std::vector<c10::IValue>&);
+using TreeUnflattenFn =
+    c10::IValue (*)(std::vector<c10::IValue>, const nlohmann::json&);
+
+using ContextLoadFn = nlohmann::json (*)(std::string_view);
```
Since the context is limited to JSON objects or strings, I'm sure this "pytree" utility cannot easily support arbitrary Python container types.

If I understand correctly, this is a C++ extension bridged with the Python pytree serialization/unserialization format for `torch.export`.

------

> But if we really want a name other than pytree, @zou3519 @zhxchen17 do you have any suggestion for a better name? maybe TreeSpec.h?

I previously tried to implement a draft for C++ pytree in `libtorch` named `torch::utils::nested`.

- #146893

FYI, TensorFlow has a module that handles nested Python containers, called [`tf.nest`](https://www.tensorflow.org/api_docs/python/tf/nest). It only has limited support with native container types in Python stdlib, and does not support registering custom types.

XuehaiPan commented on torch/nativert/detail/Pytree.cpp:
```diff
@@ -0,0 +1,485 @@
+  }
+}
+
+c10::TypePtr TreeSpec::toAtenType() const {
+  if (isLeaf()) {
+    return c10::AnyType::get();
+  } else if (uniformName_ == "builtins.tuple") {
+    std::vector<c10::TypePtr> childrenType;
+    childrenType.reserve(children_.size());
+    for (const auto& childrenSpec : children_) {
+      childrenType.emplace_back(childrenSpec.toAtenType());
+    }
+    return c10::TupleType::create(std::move(childrenType));
+  } else if (
+      uniformName_ == "builtins.list" ||
+      uniformName_ == "torch.fx.immutable_collections.immutable_list") {
+    if (children_.empty()) {
+      return c10::ListType::create(c10::AnyType::get());
+    } else {
+      return c10::ListType::create(children_[0].toAtenType());
+    }
+  } else if (
+      uniformName_ == "builtins.dict" ||
+      uniformName_ == "torch.fx.immutable_collections.immutable_dict") {
+    if (children_.empty()) {
+      return c10::DictType::create(c10::AnyType::get(), c10::AnyType::get());
+    } else {
+      return c10::DictType::create(
+          dynamicToIValue(context_[0]).type(), children_[0].toAtenType());
+    }
+  } else {
+    TORCH_CHECK(false, "Unsupported uniform name: ", uniformName());
+  }
+}
```
We should support `PyStructSequence` types in `torch.return_types.*`.

yushangdi replied:
> We should support `PyStructSequence` types in `torch.return_types.*`.

We can add this as a followup. This is mainly used for the serialized input and output specs of exported program. Currently, there's no default serialization implemented for `torch.return_types.*`. The users would need to update their _register_pytree_node call with a `serialized_type_name` kwarg for these types. If we want to hard-code the  `uniformName` here for `torch.return_types.*`, it might be good to do that along with a default serialization implementation for exported program. 

cc @zhxchen17

XuehaiPan replied:
> If we want to hard-code the `uniformName` here for `torch.return_types.*`

`torch.return_types.*` are first-party types that we can hardcode the behaviors.

yushangdi replied:
> > If we want to hard-code the `uniformName` here for `torch.return_types.*`
> 
> `torch.return_types.*` are first-party types that we can hardcode the behaviors.

@XuehaiPan yeah, I'm just saying if we hard code it here, we need to hard code it in the exported program serializer as well. And I think it might be good to add and test this in a separate PR. In this PR we're just moving the code to OSS. We can extend the code as a follow up.

yushangdi commented:
@XuehaiPan @zou3519 How about we just re-use the `NestedStructure.h` name from  https://github.com/pytorch/pytorch/pull? I can change the file name from `Pytree.h` and `Pytree.cpp` to  `NestedStructure.h` and `NestedStructure.cpp`.

XuehaiPan commented:
+1 for changing `Tree` -> `Nested`, `Leaf` -> `Atom`/`IValue`.

Another thing I want to point out is that when we call with pure-C containers, we should take care of the recursion limit. We will encounter vulnerabilities when a cycle exists in the nested structure.

```cpp
Flatten(...) {
    return FlattenHelper(..., /*depth=*/0);
}

FlattenHelper(..., size_t depth) {
    // recursion code
}
```

zou3519 reviewed (commented):
I want to bikeshed the naming more but otherwise this looks fine to me.

zhxchen17 commented on torch/nativert/detail/NestedStructure.h:
```diff
@@ -0,0 +1,178 @@
+
+void registerPytreeNode(std::string_view typeName, NodeDef nodeDef);
+
+// Serialized json tree spec should be dumped from treespec_dumps() in
+// torch.utils._pytree directly .
+NestedSpec nestedSpecLoads(
+    std::string_view json,
+    const std::vector<const Value*>& values);
+
+c10::IValue nestedUnflatten(
+    std::vector<c10::IValue> ivalues,
+    const NestedSpec& spec);
+
+std::vector<c10::IValue> nestedFlatten(
+    const c10::IValue& nested,
+    const NestedSpec& spec);
+
+std::vector<c10::IValue> nestedFlattenFromArgs(
+    const std::vector<c10::IValue>& args,
+    const std::unordered_map<std::string, c10::IValue>& kwargs,
+    const NestedSpec& spec);
```
ITree or IValueTree both sounds okay to me. I'm fine with NestedStructure as well but I think it will cause more confusion when people read about these and wonder what's the difference between this and pytree while conceptually they look pretty similar.

yushangdi replied:
@zou3519 @XuehaiPan @zhxchen17 To summarize the discussion, how about the following? 

- Use IValue/ITree/ITreeSpec name for classes/structs
- Remove nested prefix from all of the function APIs, e.g. Flatten, FlattenToTensorList...

zou3519 replied:
Using "I" as the prefix sounds good to me, this is an IValue-based flatten/unflatten API

yushangdi replied:
> Using "I" as the prefix sounds good to me, this is an IValue-based flatten/unflatten API

@zou3519 Just to double check, do you want the I-prefix to also exist on the function names? e.g. `IFlatten` or just `Flatten`, `IFlattenToTensorList` or `FlattenToTensorList`. I have a slight preference on `IFlatten` since this is less likely to have name collision (`Flatten` might be a common name).

zou3519 replied:
I don't have a preference for the function names, up to you

yushangdi commented:
@zou3519 updated to use `ITree` now :)

yushangdi committed:
[nativert] Move Pytree implementation to OSS and rename to ITree (#155136)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/155136

Torch Native Runtime RFC: https://github.com/pytorch/rfcs/pull/72

As part of the effort to open source TorchNativeRuntime (or what we call Sigmoid), we are moving the Pytree implementation to torch/:

fbcode/sigmoid/core/common -> fbcode/caffe2/torch/nativert/common

We also renamed the functions, the reason behind renaming is in this discussion: https://github.com/pytorch/pytorch/pull/155136

The functionalities and implementation remain the same.

Rename the class and functions:

1.  `TreeSpec` -> `ITreeSpec`
2.  `treeSpecLoads` -> `itreeSpecLoads`
3.  `treeFlatten` -> `itreeFlatten`
4.  `treeUnflatten` -> `itreeUnflatten`
5.  `leafApply` -> `ivalueApply`
6.  `leafApplyFromArgs` -> `ivalueApplyFromArgs`

Test Plan:
```
buck run fbcode//mode/dev-nosan  //caffe2/test/cpp/nativert:itree_test
```

OSS CI

Reviewed By: zou3519, zhxchen17

Differential Revision: D75965059
